src/pivot-table/hooks/use-selections-model.ts

```diff
@@ -48,68 +86,68 @@ export default function useSelectionsModel(
-    (qType: EngineAPI.NxSelectionCellType, qRow: number, qCol: number) => {
-      switch (qType) {
-        case NxSelectionCellType.NX_CELL_LEFT:
-          return !!selected.find(
-            (cell) =>
-              cell.qType === NxSelectionCellType.NX_CELL_TOP ||
-              (cell.qType === NxSelectionCellType.NX_CELL_LEFT && cell.qCol !== qCol),
-          );
-        case NxSelectionCellType.NX_CELL_TOP:
-          return !!selected.find(
-            (cell) =>
-              cell.qType === NxSelectionCellType.NX_CELL_LEFT ||
-              (cell.qType === NxSelectionCellType.NX_CELL_TOP && cell.qRow !== qRow),
-          );
-        default:
-          return false;
+    (cell: Cell) => {
+      if (selectedPivotCells.size === 0 || selectedRowOrColumn.selectionCellType === null) {
+        return false;
       }
+
+      if (cell.selectionCellType !== selectedRowOrColumn.selectionCellType) {
+        return true;
+      }
+
+      if (cell.selectionCellType === NxSelectionCellType.NX_CELL_LEFT) {
+        return selectedRowOrColumn.coord !== cell.x;
+      }
+
+      return selectedRowOrColumn.coord !== cell.y;
     },
-    [selected],
+    [selectedPivotCells, selectedRowOrColumn],
   );
 
   const select = useCallback(
-    (qType: EngineAPI.NxSelectionCellType, qRow: number, qCol: number) => async (evt: React.MouseEvent) => {
-      if ((evt.target as HTMLElement)?.className.includes("sn-pivot-table-column-adjuster")) {
+    (cell: Cell) => async (evt: React.MouseEvent) => {
+      if ((evt.target as HTMLElement | SVGElement)?.getAttribute("class") === "sn-pivot-table-column-adjuster") {
         return;
       }
 
       if (!selections.isActive()) {
         await selections.begin([Q_PATH]);
       }
 
-      if (isLocked(qType, qRow, qCol)) {
+      if (isLocked(cell)) {
         return;
       }
 
-      setSelected((prev) => {
-        const values = [...prev];
-        const idx = values.findIndex((cell) => cell.qType === qType && cell.qRow === qRow && cell.qCol === qCol);
-        if (idx > -1) {
-          values.splice(idx, 1);
-        } else {
-          values.push({ qType, qRow, qCol });
-        }
-
-        selections
-          .select({
-            method: "selectPivotCells",
-            params: [Q_PATH, values],
-          })
-          .catch(() => {});
-
-        return values;
-      });
+      const { nextSelectedPivotCells, nextSelectedRowOrColumn } = getNextState(
+        cell,
+        selectedPivotCells,
+        selectedRowOrColumn,
+      );
+
+      try {
+        await selections.select({
+          method: "selectPivotCells",
+          params: [
+            Q_PATH,
+            Array.from(nextSelectedPivotCells.values()).map((c) => ({
+              qType: c.selectionCellType,
+              qRow: c.y,
+              qCol: c.x,
+            })),
+          ],
+        });
+
+        setSelectedPivotCells(nextSelectedPivotCells);
+        setSelectedRowOrColumn(nextSelectedRowOrColumn);
```

Comment:
~~doesn't these two line replace the same value to states?
because not transformation happens to values of: `nextSelectedPivotCells` and `nextSelectedRowOrColumn` since they been retrieved from `getNextState()` on line 121~~

isn't it better to keep state update and data transformation at the same place? (just to not fall into 👆 trap while reading the code)

Reply:
We first need to transform the data, send it to Engine and if all goes well we can update the state. So I don't think there is a way to keep "state update" and "data transformation" in the same place?